config/config.go

```diff
@@ -0,0 +1,110 @@
+package config
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/knq/ini"
+	"github.com/knq/ini/parser"
+)
+
+type KV struct {
+	Key   string
+	Value string
+}
+
+type Generator struct {
+	Command string
+	Params  []KV
+}
+
+type Config struct {
+	Dir        string
+	Patterns   []string
+	Generators []Generator
+}
+
+func Load(path string) (*Config, error) {
+	origPath := path
+	var err error
+	if path == "" {
+		path, err = os.Getwd()
+		if err != nil {
+			return nil, fmt.Errorf("error getting working directory: %v", err)
+		}
+	}
+
+	var cfg *Config
+	for {
+		configPath := filepath.Join(path, ".gunkconfig")
+		reader, err := os.Open(configPath)
+		if err != nil {
+			path = filepath.Dir(path)
+			if path == "/" {
+				return nil, fmt.Errorf("could not find a .gunkconfig")
+			}
+			continue
+		}
+		defer reader.Close()
+		cfg, err = load(reader)
+		if err != nil {
+			return nil, fmt.Errorf("error loading %q: %v", configPath, err)
+		}
+		if cfg != nil {
```

Comment:
why would `load` ever return nil error and nil cfg? this case could also lead to weird behavior, such as parsing multiple config files and leaving them open (see the `defer`).